src/Npgsql/ConnectorPool.cs

```diff
@@ -413,25 +382,12 @@ internal void Release(NpgsqlConnector connector)
 
             connector.Reset();
 
-            var sw = new SpinWait();
-
             while (true)
             {
-                var state = State.Copy();
-
                 // If there are any pending open attempts in progress hand the connector off to them directly.
-                // Note that in this case, state changes (i.e. decrementing State.Waiting) happens at the allocating
-                // side.
-                if (state.Waiting > 0)
+                // Note that in this case, state changes happens at the allocating side.
+                if (_waiting.TryDequeue(out var waitingOpenAttempt))
                 {
-                    if (!_waiting.TryDequeue(out var waitingOpenAttempt))
-                    {
-                        // _waitingCount has been increased, but there's nothing in the queue yet - someone is in the
-                        // process of enqueuing an open attempt. Wait and retry.
-                        sw.SpinOnce();
-                        continue;
-                    }
-
                     var tcs = waitingOpenAttempt.TaskCompletionSource;
 
                     // We have a pending open attempt. "Complete" it, handing off the connector.
```

Comment:
Waiter could enqueue and read Idle == 0 right between these two ops, will think of a good way to deal with that